packages/router/router.ts

```diff
@@ -1589,9 +1578,41 @@ export function createRouter(init: RouterInit): Router {
     let redirectHistoryAction =
       replace === true ? HistoryAction.Replace : HistoryAction.Push;
 
-    await startNavigation(redirectHistoryAction, navigation.location, {
-      overrideNavigation: navigation,
-    });
+    let { formMethod, formAction, formEncType, formData } = state.navigation;
+
+    // If this was a 307/308 submission we want to preserve the HTTP method and
+    // re-submit the POST/PUT/PATCH/DELETE as a submission navigation to the
+    // redirected location
+    if (
+      [307, 308].includes(redirect.status) &&
+      formMethod &&
+      formMethod !== "get" &&
+      formMethod !== "head" &&
```

Comment:
Since we're already using `includes` above, we can maybe do

```suggestion
      !["get", "head"].includes(formMethod) &&
```